surmise v0.2 staging

mosesyhc opened the pull request:
*This has fixed CI, and is branched from latest develop.*

All methods that goes into v0.2 (TBD) can be merged here, including:

- [x] #52 
- [x] #54 
- [x] #55 
- ~~Miscellaneous~~

wildsm commented on docs/examples.rst:
```diff
@@ -37,11 +37,11 @@ Recovered (SIR) type epidemic model via ``surmise``'s  ``emulator`` and ``calibr
-`Link to Example 4 <https://nbviewer.jupyter.org/github/surmising/surmise/blob/master/examples/Example4/Example4_nb.ipynb>`_.
+`Link to Example 4 <https://nbviewer.jupyter.org/github/surmising/surmise/blob/main/examples/Example4/Example4_nb.ipynb>`_.
 
 Example 5
```
TODO: Revisit Example 5 after GPy branch merged

wildsm replied:
done

mosesyhc commented:
Saving this here for removing all .DS_Store

```
find . -name .DS_Store -print0 | xargs -0 git rm -f --ignore-unmatch
```

wildsm commented on CHANGELOG.rst:
revisit before merge/moving PR from draft

wildsm commented:
WIP/Status update - I'm about halfway through the files so far in preparation for this PR.
Notes:
- See #67 
- I am assuming that version 0.2 will be packaged from the surmising org, if @MattPlumlee and @ozgesurer want to move to bandframework org before version 0.2, we can easily migrate this repo soon.
- I am not doing full checks of `.ipynb` files and functionality
-

mosesyhc commented:
- I am confirming that the `.ipynb`'s run without issues.  
  - The last two calibration runs in Example4_nb take significant amount of time.
  - For v0.2, no content will be changed/updated in the Examples.

ozgesurer commented:
I am ok if we want to move to bandframework org before version 0.2.

mosesyhc commented:
@ozgesurer, @wildsm: 
I believe we have reviewed most if not all updates required for v0.2 release, with these last bits as follows:

- [ ] #72 
- [ ] Decision to stay in surmising vs. migrate to bandframework
  - [ ] Follow up to update submodule pointer, OR
  - [ ] Follow up to edit documentation and SDK that refers to "surmising" organization

Post-merge TODOs:
- [ ] Update git tag release, upload wheel and tar.gz
- [ ] Verify readthedocs documentation
- [ ] Verify CHANGELOG
- [ ] Require BAND members to review
- [ ] Bring develop branch up to date (or rename this to develop)
- [ ] Remove this branch

wildsm reviewed (approved):
Looks great, Good checklist (I added an item) in the conversation.
Still potentially pending the pypi workflow merge.

mosesyhc commented:
> Looks great, Good checklist (I added an item) in the conversation. Still potentially pending the pypi workflow merge.

I agree to hold the pypi publishing workflow.  It does not have to rush.

ozgesurer commented:
@mosesyhc Thanks for the great work! I will write the steps that I follow on my Mac:

Create a virtual environment via:

```
python3 -m venv venv/ 
source venv/bin/activate
```

Install the requirements:

`pip install -r requirements.txt`

Install surmise:

`pip install -e .`

Check installation:

`pip show surmise
`

This is the print (I don't understand why the version number is 0.1.2):

Name: surmise
Version: 0.1.2.dev444+g36a20dd
Summary: A modular interface for surrogate models and tools
Home-page: 
Author: Matthew Plumlee, Stefan M. Wild
Author-email: Özge Sürer <surero@miamioh.edu>, "Moses Y-H. Chan" <mosesyhc@u.northwestern.edu>
License: MIT License
        
        Copyright (c) 2023 The Surmise Development Team
        
        Permission is hereby granted, free of charge, to any person obtaining a copy
        of this software and associated documentation files (the "Software"), to deal
        in the Software without restriction, including without limitation the rights
        to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
        copies of the Software, and to permit persons to whom the Software is
        furnished to do so, subject to the following conditions:
        
        The above copyright notice and this permission notice shall be included in all
        copies or substantial portions of the Software.
        
        THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
        IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
        FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
        AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
        LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
        OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
        SOFTWARE.
        
Location: 
Requires: numpy, scipy, scikit-learn
Required-by:

mosesyhc commented:
@ozgesurer: 
It looks like the installation is successful, correct?

The version number is automatically generated by git_scm (see 791d785).  What it does is that it fetches the most recent tag and check if the software has changed since then.  Since we have updated the software, it thinks that we have moved forward from v0.1.1 but still in development, therefore the v0.1.2dev+

Once we create a new tag, say v0.2.0, it will check and generate the correct version.  

I can give it a try and let you know how it would turn out.

ozgesurer commented:
Tes: If I install surmise as above, `python setup.py test` did not work. However, if I do `./run-tests.sh` from the `tests/` directory, all tests are successfully run.

ozgesurer commented:
> @ozgesurer: It looks like the installation is successful, correct?
> 
> The version number is automatically generated by setuptools_scm (see [791d785](https://github.com/surmising/surmise/commit/791d7852ff66f7f8665479e7f7f334a03e23d7fc)). What it does is that it fetches the most recent tag and check if the software has changed since then. Since we have updated the software, it thinks that we have moved forward from v0.1.1 but still in development, therefore the v0.1.2dev+
> 
> Once we create a new tag, say v0.2.0, it will check and generate the correct version.
> 
> I can give it a try and let you know how it would turn out.

Sounds good on version number if it depends on our tag. And, yes the installation is succesful.

mosesyhc commented:
> > @ozgesurer: It looks like the installation is successful, correct?
> > The version number is automatically generated by setuptools_scm (see [791d785](https://github.com/surmising/surmise/commit/791d7852ff66f7f8665479e7f7f334a03e23d7fc)). What it does is that it fetches the most recent tag and check if the software has changed since then. Since we have updated the software, it thinks that we have moved forward from v0.1.1 but still in development, therefore the v0.1.2dev+
> > Once we create a new tag, say v0.2.0, it will check and generate the correct version.
> > I can give it a try and let you know how it would turn out.
> 
> Sounds good on version number if it depends on our tag. And, yes the installation is succesful.

I just pushed a v0.2.0 tag. Try to reinstall surmise and see if the correct version is installed.

ozgesurer commented:
> @mosesyhc Tests: If I install surmise as above, `python setup.py test` did not work. However, if I do `./run-tests.sh` from the `tests/` directory, all tests are successfully run.

I believe the reason why `python setup.py test` did not work is because of the way how I installed.

mosesyhc commented:
> > @mosesyhc Tests: If I install surmise as above, `python setup.py test` did not work. However, if I do `./run-tests.sh` from the `tests/` directory, all tests are successfully run.
> 
> I believe the reason why `python setup.py test` did not work is because of the way how I installed.

It might be a little more nuanced, because I moved part of our setup script into pyproject.toml (check the discussion on https://github.com/surmising/surmise/commits/v0.2-staging/pyproject.toml

I am double-checking how tests are run with setup.py

ozgesurer commented:
> > > @ozgesurer: It looks like the installation is successful, correct?
> > > The version number is automatically generated by setuptools_scm (see [791d785](https://github.com/surmising/surmise/commit/791d7852ff66f7f8665479e7f7f334a03e23d7fc)). What it does is that it fetches the most recent tag and check if the software has changed since then. Since we have updated the software, it thinks that we have moved forward from v0.1.1 but still in development, therefore the v0.1.2dev+
> > > Once we create a new tag, say v0.2.0, it will check and generate the correct version.
> > > I can give it a try and let you know how it would turn out.
> > 
> > 
> > Sounds good on version number if it depends on our tag. And, yes the installation is succesful.
> 
> I just pushed a v0.2.0 tag. Try to reinstall surmise and see if the correct version is installed.

Yep, it worked! See below:

Name: surmise
Version: 0.2.0
Summary: A modular interface for surrogate models and tools
Home-page: 
Author: Matthew Plumlee, Stefan M. Wild
Author-email: Özge Sürer <surero@miamioh.edu>, "Moses Y-H. Chan" <mosesyhc@u.northwestern.edu>
License: MIT License
        
        Copyright (c) 2023 The Surmise Development Team
        
        Permission is hereby granted, free of charge, to any person obtaining a copy
        of this software and associated documentation files (the "Software"), to deal
        in the Software without restriction, including without limitation the rights
        to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
        copies of the Software, and to permit persons to whom the Software is
        furnished to do so, subject to the following conditions:
        
        The above copyright notice and this permission notice shall be included in all
        copies or substantial portions of the Software.
        
        THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
        IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
        FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
        AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
        LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
        OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
        SOFTWARE.
        
Location: /Users/ozgesurer/Desktop/GithubRepos/surmise
Requires: numpy, scipy, scikit-learn
Required-by:

mosesyhc commented:
> > > @mosesyhc Tests: If I install surmise as above, `python setup.py test` did not work. However, if I do `./run-tests.sh` from the `tests/` directory, all tests are successfully run.
> > 
> > 
> > I believe the reason why `python setup.py test` did not work is because of the way how I installed.
> 
> It might be a little more nuanced, because I moved part of our setup script into pyproject.toml (check the discussion on https://github.com/surmising/surmise/commits/v0.2-staging/pyproject.toml
> 
> I am double-checking how tests are run with setup.py

Several observations:
- it looks like `python setup.py test` is deprecated and will be removed from setuptools. (https://setuptools.pypa.io/en/latest/deprecated/commands.html#test-build-package-and-run-a-unittest-suite)
- Our `./run-tests.sh` selects only some tests that are written inside `tests/` directory.
- To run all written tests, it is possible to simply run `pytest` within the root directory of surmise.  Some tests will fail as of now, because they are looking for features that are not implemented.

I think we need to decide if running `./run-tests.sh` is currently sufficient.  If we want to include `pytest`, then we should review the currently failing tests.

In the long run, we probably should update some of the tests to test *all* methods.. but maybe this is too hefty a task for v0.2.

ozgesurer commented:
> > > > @mosesyhc Tests: If I install surmise as above, `python setup.py test` did not work. However, if I do `./run-tests.sh` from the `tests/` directory, all tests are successfully run.
> > > 
> > > 
> > > I believe the reason why `python setup.py test` did not work is because of the way how I installed.
> > 
> > 
> > It might be a little more nuanced, because I moved part of our setup script into pyproject.toml (check the discussion on https://github.com/surmising/surmise/commits/v0.2-staging/pyproject.toml
> > I am double-checking how tests are run with setup.py
> 
> Several observations:
> 
> * it looks like `python setup.py test` is deprecated and will be removed from setuptools. (https://setuptools.pypa.io/en/latest/deprecated/commands.html#test-build-package-and-run-a-unittest-suite)
> * Our `./run-tests.sh` selects only some tests that are written inside `tests/` directory.
> * To run all written tests, it is possible to simply run `pytest` within the root directory of surmise.  Some tests will fail as of now, because they are looking for features that are not implemented.
> 
> I think we need to decide if running `./run-tests.sh` is currently sufficient. If we want to include `pytest`, then we should review the currently failing tests.
> 
> In the long run, we probably should update some of the tests to test _all_ methods.. but maybe this is too hefty a task for v0.2.

I think `./run-tests.sh` is sufficient for now. Actions also use this way.

mosesyhc commented:
> > > > > @mosesyhc Tests: If I install surmise as above, `python setup.py test` did not work. However, if I do `./run-tests.sh` from the `tests/` directory, all tests are successfully run.
> > > > 
> > > > 
> > > > I believe the reason why `python setup.py test` did not work is because of the way how I installed.
> > > 
> > > 
> > > It might be a little more nuanced, because I moved part of our setup script into pyproject.toml (check the discussion on https://github.com/surmising/surmise/commits/v0.2-staging/pyproject.toml
> > > I am double-checking how tests are run with setup.py
> > 
> > 
> > Several observations:
> > 
> > * it looks like `python setup.py test` is deprecated and will be removed from setuptools. (https://setuptools.pypa.io/en/latest/deprecated/commands.html#test-build-package-and-run-a-unittest-suite)
> > * Our `./run-tests.sh` selects only some tests that are written inside `tests/` directory.
> > * To run all written tests, it is possible to simply run `pytest` within the root directory of surmise.  Some tests will fail as of now, because they are looking for features that are not implemented.
> > 
> > I think we need to decide if running `./run-tests.sh` is currently sufficient. If we want to include `pytest`, then we should review the currently failing tests.
> > In the long run, we probably should update some of the tests to test _all_ methods.. but maybe this is too hefty a task for v0.2.
> 
> I think `./run-tests.sh` is sufficient for now. Actions also use this way.

In that case, let's remove the instruction from README.

ozgesurer commented:
Documentation: `make html` works fine including use cases with reference list.

wildsm commented:
Regarding the tests and pytest, I think this good to have an open issue on and to note in the release/changelog. Agreed that `run-tests.sh` sufficient for v0.2

mosesyhc commented:
> Documentation: `make html` works fine including use cases with reference list.

And the most recent readthedocs build passed!

mosesyhc commented:
@ozgesurer, @wildsm: If we could hold off merging this to main by another day or two, there might be a minor update for `nuclear-ROSE` emulator incoming..

ozgesurer commented:
> @ozgesurer, @wildsm: If we could hold off merging this to main by another day or two, there might be a minor update for `nuclear-ROSE` emulator incoming..

Fine with me.

ozgesurer commented:
> > Documentation: `make html` works fine including use cases with reference list.
> 
> And the most recent readthedocs build passed!

Nice!! I think everything looks very good.

wildsm commented:
All authors have now agreed to move the org (already done). 

Moving forward with the checklist #76 

But this also means that we've done almost all that is needed to complete v2 staging

wildsm commented:
Promoting earlier post-merge TODOs:

https://github.com/bandframework/surmise/pull/53#issuecomment-1716401441

mosesyhc commented:
@wildsm, @ozgesurer: It seems like we have completed all tasks for merging to main.  
The post-merge comment has been repeated in a new issue #82.

wildsm reviewed (approved):
Thanks Ozge and Moses!
